whole-tale/girder_wholetale
# Added cullIdleInstances handler

Fixes https://github.com/whole-tale/girder_wholetale/issues/477#

**Test case**
* Requires https://github.com/whole-tale/girder/pull/11 or `wholetale/girder:heartbeat` image based on that PR
* Requires https://github.com/whole-tale/deploy-dev/pull/49
* Using `girder-shell`, temporarily set the `idleTimeout` on the JupyterLab image to 1 minute
```
    docker exec -it <girder> bash
    $ girder-shell

    from girder.plugins.wholetale.models.image import Image
    image = Image().findOne({"name": "JupyterLab"})
    image["idleTimeout"] = 1
    Image().save(image)
```
* Create and run a JupyterLab tale
* Optionally tail girder logs (`docker exec -it <girder> tail -f  tail -f /home/girder/.girder/logs/info.log`)
   * `Culling idle instances message should appear every `heartbeat` seconds (5 min default)
* With IDE open in either iframe or popout, wait for 5 minutes
   * Confirm that instance is not stopped
* Select new tale tab (e.g., Metadata)  or browser tab, wait for 5 minutes
   * If tailing logs, note `Stopping instance X: idle timeout exceeded`
   * Confirm that instance is stopped
* Go to https://tmp-X.local.wholetale.org 
   * Confirm that custom 404 page is displayed
